MAINT: Implement float128 dragon4 for IBM double-double (ppc64)

ahaldane opened the pull request:
This PR reorganizes the dragon4 code to make all the different float formats more explicit, and then implements the IBM-double-double Dragon4 method.

I might want to put some more tests, but putting it up now to run the github tests. It passes tests on ppc64, python2/3.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -1076,6 +1076,37 @@ def format_float_positional(x, precision=None, unique=True,
+    format_float_scientific
+    format_float_positional
+    """
+    return printf_float_g(val, precision)
```
Why the python wrapper?

ahaldane replied:
Just for the docstring

eric-wieser replied:
Precedent seems to be to use `add_new_docs` for that.

eric-wieser commented on numpy/core/src/multiarray/dragon4.c:
```diff
@@ -640,8 +699,8 @@ static void
-    BigInt temp1;
-    BigInt temp2;
+    static BigInt temp1;
+    static BigInt temp2;
```
What gain is there to making these static?

ahaldane replied:
Maybe this is a bad idea... but I was worried about using up too much stack space. Each BigInt is 4k.

ahaldane replied:
(I didn't want to use stack beause these are big variables, and I didn't want to use the heap cause it's slower. This way is uses the "data" memory.)

eric-wieser replied:
Do these actually need to be big, or are some of these bigints only int128s? Could you use `alloca` to allocate bigints only of the size needed on the stack?

ahaldane replied:
The BigInts have to be able to grow as the computation progresses. Most of the time the BigInt needs much less than 4k, but sometimes (very hard to know beforehand) it needs a lot more. I think in all cases they can potentially be more than int128.

By the way, there may be some optimizations to make here. Ryan Jucket optimized his BigInts for speed, but his were only 140 bytes and he saved them on the stack, because he only did up to 64 bit floats. But for float128 I need much bigger ints. 

For float64, the max exponent (base 2) is 1023, so we need BigInts up to 1023 bits, plus 53 for the mantissa. (140 bytes = 1120 bits). For float128, the max exponent (base 2) is 16383, so we need 16383 bits plus 64 for the mantissa, or about 2kb. The 4k I used is a bit overkill.

charris replied:
Is it thread safe? Putting things on the stack has that advantage.

ahaldane replied:
No, it isn't. I'm still hesitant about the stack.. but let me try mallocing them on the heap, and see if there's a performance hit.

From some quick performance tests I did a while ago, it seemed possible there is some python/numpy overhead to printing that might dwarf the cost of Dragon4 itself, in which case the speed of the BigInt allocation might be negligible.

ahaldane replied:
I did some measurement:

For this PR currently, using "data" memory, I get that float64 printing with Dragon4 is about 5% slower that `printf("%.18g"`, when printing numbers generated by `np.random.rand`.

When I switch to malloc'ing on the heap, there is a pretty big slowdown, so that Dragon4 becomes about 300% slower than printf (i.e., a factor of 3). I tried to optimize by malloc'in all the needed BigInts at once to get a single malloc.

Is it a big deal if the repr is not re-entrant? We don't release the GIL for printing anywhere, so I would favor keeping the `static`.

We can also consider leaving the BigInts in the stack space: The total space needed by a single call to Dragon4 is about 24kb (6 BigInts). Typical stacks are 8mb (linux) and 1mb (windows). The goal of putting this into stack instead of data is to make Dragon4 re-entrant, but with such a large stack increase there is a real danger of using up the stack when re-entering. So even if we allocate in stack, we should recommend *not* to re-enter. So might as well allocate in "data".

ahaldane commented:
If someone has a "darwin" system, would be nice to run tests. I got rid of a bunch of darwin hardcoded macro stuff, which I think I fixed, but I'm not totally sure.

Also, I just remembered: I still need to add some macro magic to stop gcc complaining that some of the formatting functions are unused on some systems.

eric-wieser commented on numpy/core/src/multiarray/dragon4.c:
```diff
@@ -1028,18 +1087,20 @@ Dragon4(const npy_uint64 mantissa, const npy_int32 exponent,
-    BigInt optionalMarginHigh;
+    static BigInt optionalMarginHigh;
 
     const npy_float64 log10_2 = 0.30102999566398119521373889472449;
```
Does this need more digits for IBM double double?

ahaldane replied:
no, the exact value here isn't super important, it just optimizes some branch prediction.

ahaldane replied:
This value is Discussed on Ryan Juckett's website under "logarithms", plus in the Burger & Dybvig paper.

http://www.ryanjuckett.com/programming/printing-floating-point-numbers/

eric-wieser replied:
Presumably having more digits would further optimize the prediction though? Or is this already more digits than an IBM double double can hold?

ahaldane replied:
Well, in some sense the main innovation of Dragon4 and the Burger & Dybvig paper is the fact that this calculation uses a simple float64 multiplication! And the calculation is already inexact enough that these extra digits shouldn't matter.

This value of `log10_2` is used in only one place, in the calculation of `digitExponent`. Note there is a long comment there (not by me) partly explaining this calculation, and you can also read the intro to the Burger & Dybvig which explains the motivation, and section 3.2 is entirely dedicated to exlaining it. The goal is to calculate a the base-10 exponent. Older algorithms (Steel and White) used a very expensive BigInt calculation to do this, which dominated the cost of the algorithm. Burger & Dybvig's contribution was to realize that if you just use a simple float64 calculation, you would at most be off by 1, and you could later detect this and correct it. That's why there's two branches: A "correct by one" path and a "no correction" path. Ryan Jucket realized that by adding an additional constant you could get it to go into the faster path more often, without risk of being off by more than one.

Just to try to illustrate, here is the actual calculation:
```
digitExponent = ceil( (mantissa - exponent)*log10_2 + 0.69)
```
Ryan Juckett added that extra 0.69, noting that it "moves the error range from about (-0.301, 0.0] to about (-0.991, -0.69]". We want to move the lower bound as close to -1 as possible, so in principle we could even use 0.699. But there is some danger getting too close to -1 since there is error in the float64 multiply with `log10_2` itself, B&D estimate it is up to `10**-10` for float64 mantissas and I estimate it is `10**-8` for float80 mantissas.

In summary, the computation is already off by about 0.009, meaning branch prediciton could be optimized by this amount, but that's so small it doesn't seem worth touching.

eric-wieser replied:
My misunderstanding here was not noticing that this was a float64 instead of a longdouble - but thanks for the explanation anyway!

eric-wieser commented on numpy/core/setup_common.py:
```diff
@@ -333,9 +333,9 @@ def _pyod3():
+_IBM_DOUBLE_DOUBLE_BE = (['301', '235', '157', '064', '124', '000', '000', '000'] +
                      ['000'] * 8)
-_DOUBLE_DOUBLE_LE = (['000', '000', '000', '124', '064', '157', '235', '301'] +
+_IBM_DOUBLE_DOUBLE_LE = (['000', '000', '000', '124', '064', '157', '235', '301'] +
```
This looks quite similar to tan approach used in `finfo`, and it might be nice to combine them somehow.

eric-wieser replied:
Far more alarmingly, this section is identical to `numpy\_build_utils\common.py`

eric-wieser commented on numpy/core/src/multiarray/dragon4.c:
```diff
@@ -2424,14 +2450,14 @@ Dragon4_PrintFloat128(char *buffer, npy_uint32 bufferSize, FloatVal128 value,
 
     /* deconstruct the floating point value */
-    floatExponent = GetExponent_F128(&value);
-    floatMantissa = GetMantissa_F128(&value);
```
Might be convenient to add a
```
/* Create a mask with the lowest n bits set */
uint64_t mask_lowest_64(int32_t n) {
    return ~(~0 << n);
}
```
function, which saves from having to count the `F`s

eric-wieser commented on numpy/add_newdocs.py:
```diff
@@ -5525,6 +5525,37 @@ def luf(lamdaexpr, *args, **kwargs):
+        printf("%.*g", precision, val);
+
+    for half/float/double, or replacing 'g' by 'Lg' for longdouble. This
+    method is designed to help cross-validate the format_float_* methods.
```
Should this maybe go in the multiarray_tests.c.src instead (and not be exposed at all?)

ahaldane replied:
I'd like to expose it in case it helps users debug any problems.

eric-wieser replied:
I mispoke - it would still be exposed through `numpy.core.multiarray_tests`

ahaldane replied:
Fair enough, will move.

eric-wieser commented on numpy/core/setup.py:
```diff
@@ -452,17 +452,16 @@ def generate_config_h(ext, build_dir):
+                       'MOTOROLA_EXTENDED_12_BYTES_BE',
+                       'IEEE_QUAD_LE', 'IEEE_QUAD_BE',
+                       'IEEE_DOUBLE_LE', 'IEEE_DOUBLE_BE',
+                       'IBM_DOUBLE_DOUBLE_BE', 'DOUBLE_DOUBLE_LE']:
```
This check seems pointless, because `long_double_representation` always returns a member from this list or throws an exception.

ahaldane replied:
true, removed

eric-wieser commented on numpy/core/src/npymath/npy_math_private.h:
```diff
@@ -477,7 +476,7 @@ do {                                                            \
      (((IEEEl2bitsrep_part)(v) << LDBL_MANH_SHIFT) & LDBL_MANH_MASK))
 
-#endif /* #ifndef HAVE_LDOUBLE_DOUBLE_DOUBLE_BE */
+#endif /* #ifndef HAVE_IBM_LDOUBLE_DOUBLE_DOUBLE_BE */
```
Comment is wrong

eric-wieser commented on numpy/core/tests/test_scalarprint.py:
```diff
@@ -199,6 +200,61 @@ def test_dragon4_interface(self):
+        # versions of glibc have a bug in strtold that they fail to get
+        # float128 values from a string if in the subnormal range.
+        x = np.float128('2.123123123123123123123123123123123e-286')
+        got = [str(x/np.float128('2e' + str(i))) for i in range(0,40)]
```
Does `float128(2**i)` work here, or did I never finish implementing that?

ahaldane replied:
No, not in master. 

Also, the comment is wrong: strtold is giving the right result, and also correctly setting errno. I'll just remove the comment.

eric-wieser commented on numpy/core/src/multiarray/dragon4.c:
```diff
@@ -2229,20 +2245,20 @@ Dragon4_PrintFloat32(char *buffer, npy_uint32 bufferSize, npy_float32 value,
 
     /* output the sign */
-    if (IsNegative_F32(&floatUnion)) {
+    if ((floatUnion.integer >> 31) != 0) {
```
For clarity, `bool floatSign = floatUnion.integer >> 31`  would match the above mantissa and exponent nicely

eric-wieser commented on numpy/core/src/multiarray/dragon4.c:
```diff
@@ -2384,33 +2401,51 @@ Dragon4_PrintFloat64(char *buffer, npy_uint32 bufferSize, npy_float64 value,
+ *
+ * sign:      1 bit,  second u64
+ * exponent: 15 bits, second u64
+ * intbit     1 bit,  first u64
```
???

ahaldane replied:
This is the "integer part" bit, which is always 1 for normal numbers. Since it never changes, it is an "implicit" bit in the binary64 format, but for some reason they kept it explicit in the 80-bit extended format.

It can be 0 for some "Pseudo" nan/inf/0 values, which I have ignored completely because modern processors do not generate them.

charris commented:
Leaving this to 1.15, too big for a backport and is really an enhancement.

ahaldane commented:
OK

eric-wieser commented on numpy/core/include/numpy/npy_common.h:
```diff
@@ -1002,6 +1011,7 @@ typedef struct { npy_longdouble real, imag; } npy_clongdouble;
 #    define PyComplex256ArrType_Type PyCFloatArrType_Type
 #define NPY_FLOAT128_FMT NPY_FLOAT_FMT
 #define NPY_COMPLEX256_FMT NPY_CFLOAT_FMT
+#define NPY_FLOAT_BINFMT Intel_extended128
```
These names are very confusing, because `NPY_FLOAT128_FMT` is a string for format strings, yet `NPY_FLOAT_BINFMT` is a preprocessor token that isn't valid unless concatenated

ahaldane replied:
Yeah, I could call it `NPY_FLOAT_BINFMT_NAME` or similar to lessen the confusion.

eric-wieser replied:
Perhaps it should be `NPY_LONGDOUBLE_FMT_NAME`?

eric-wieser commented on numpy/core/src/multiarray/multiarray_tests.c.src:
```diff
@@ -1729,6 +1731,60 @@ call_npy_@name@@suffix@(PyObject *NPY_UNUSED(self), PyObject *args)
+
+    if (PyArray_IsScalar(obj, Half)) {
+        npy_half x = ((PyHalfScalarObject *)obj)->obval;
+        PyOS_snprintf(str, 1024, "%.*g", precision, npy_half_to_double(x));
```
nit: `1024` &rarr; `sizeof(str)`

eric-wieser commented on numpy/core/src/multiarray/multiarray_tests.c.src:
```diff
@@ -1729,6 +1731,60 @@ call_npy_@name@@suffix@(PyObject *NPY_UNUSED(self), PyObject *args)
+printf_float_g(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds)
+{
+    PyObject *obj;
+    int precision=-1;
```
No point initializing here

eric-wieser reviewed (commented):
`dragon4.c` is getting very long. Can we split it up in any way? It seems like the higher level `Dragon4_PrintFloat_IBM_double_double` function and similar could be in a different file from the lower level `Format_floatbits` and everything else.

charris commented:
Needs rebase, also  `multiarray_tests.c.src` has been renamed to `_multiarray_tests.c.src`.

We should finish this up. I'd be inclined to malloc the arrays on the heap and worry about speed if it becomes a problem, at which point one could consider other solutions such a pool. Having global variables can lead to bugs that are hard to track down.

charris commented:
A release note would also be good.

charris commented:
Might also be good to have a benchmark if execution time is a concern.

ahaldane committed:
BUG: Implement float128 dragon4 for IBM double-double (ppc64)

ahaldane commented:
Rebased and updated.

I removed all the static variables and replaced them by a dummy BigInt-manager, which returns some fast data-allocated BigInts the first time dragon4 is called, and fails for now if dragon4 is re-entered. 

We can upgrade it later if we want dragon4 to support threading. If you push me, I could try to figure out how to use `PyThread_acquire_lock` (is that available on all platforms?), and actually implement a re-entrant version, but I'll leave the simpler version here today.

eric-wieser commented on numpy/core/setup_common.py:
```diff
@@ -364,22 +364,28 @@ def long_double_representation(lines):
                 saw = copy.copy(read)
+                # if the content was 12 bytes, we only have 32 - 8 - 12 = 12
+                # "before" bytes. In other words the first 4 "before" bytes went
+                # past the sliding window.
```
This is possibly easier to express as:
```
if read[-8:] == _AFTER_SEQ:
    rest = read[:-8]  # before and match
    if rest[-12:] == _BEFORE_SEQ[-12:]:
        match = rest[:12]
        if match == _INTEL_EXTENDED_12B:
             ...
```

eric-wieser commented on numpy/core/src/multiarray/_multiarray_tests.c.src:
```diff
@@ -1828,6 +1830,61 @@ call_npy_@name@@suffix@(PyObject *NPY_UNUSED(self), PyObject *args)
 /**end repeat**/
 
+/*
+ * For development/testing purposes, it's conventient to have access to the
```
typo: convenient

eric-wieser commented on numpy/core/src/multiarray/_multiarray_tests.c.src:
```diff
@@ -1828,6 +1830,61 @@ call_npy_@name@@suffix@(PyObject *NPY_UNUSED(self), PyObject *args)
+    }
+    else{
+        double val = PyFloat_AsDouble(obj);
+        if (PyErr_Occurred()) {
```
Think we'd usually use `error_converting(val)` here

eric-wieser commented on numpy/core/src/multiarray/_multiarray_tests.c.src:
```diff
@@ -1828,6 +1830,61 @@ call_npy_@name@@suffix@(PyObject *NPY_UNUSED(self), PyObject *args)
+    }
+    else if (PyArray_IsScalar(obj, Double)) {
+        npy_double x = ((PyDoubleScalarObject *)obj)->obval;
+        PyOS_snprintf(str, sizeof(str), "%.*g", precision, x);
```
Would be nice to use `lg` here, but that's not supported in C90 :(

eric-wieser replied:
Might be nice to add a comment to that effect, actually

ahaldane replied:
I see some comments online that it `lg` is more consistent with scanf, is that why?

eric-wieser replied:
Yeah, that's why. `va_list` doesn't support `float` which is why it doesn't actually matter for printf (but it does support `float *`, which is why it matters for scanf)

eric-wieser commented on numpy/core/src/multiarray/_multiarray_tests.c.src:
```diff
@@ -1828,6 +1830,61 @@ call_npy_@name@@suffix@(PyObject *NPY_UNUSED(self), PyObject *args)
+    PyObject *obj;
+    int precision;
+
+    if (!PyArg_ParseTuple(args,"Oi:printf_float_g", &obj, &precision)) {
```
Shouldn't this name be the same as the exposed python function, `format_float_OSprintf_g`?

eric-wieser commented on numpy/core/src/multiarray/dragon4.h:
```diff
@@ -39,6 +39,33 @@
+
+#define NPY_HALF_BINFMT_NAME IEEE_binary16
+
+/* NPY_FLOAT_BINFMT_NAME and NPY_DOUBLE_BINFMT defined in npy_common.h */
```
Do you mean `NPY_DOUBLE_BINFMT_NAME` here?

eric-wieser commented on numpy/core/src/multiarray/dragon4.h:
```diff
@@ -64,15 +91,22 @@ typedef enum TrimMode
+                   int sign, TrimMode trim, int pad_left, int pad_right);\
+PyObject *\
+Dragon4_Scientific_##Type(npy_type *val, DigitMode digit_mode, int precision,\
+                   int sign, TrimMode trim, int pad_left, int exp_digits);
```
nit: might be clearer to indent all of this macro.

eric-wieser commented on numpy/core/src/multiarray/dragon4.h:
```diff
@@ -64,15 +91,22 @@ typedef enum TrimMode
-Dragon4_Positional_AnySize(void *val, size_t size, DigitMode digit_mode,
-                           CutoffMode cutoff_mode, int precision, int sign,
-                           TrimMode trim, int pad_left, int pad_right);
+#define make_dragon4_typedecl(Type, npy_type, format) \
```
Why is this `format` argument here at all?

eric-wieser commented on numpy/core/src/private/npy_fpmath.h:
```diff
@@ -7,45 +7,24 @@
 
-#ifdef NPY_OS_DARWIN
-    /* This hardcoded logic is fragile, but universal builds makes it
-     * difficult to detect arch-specific features */
```
This was replaced by the python-side stuff?

eric-wieser commented on numpy/core/tests/test_scalarprint.py:
```diff
@@ -250,6 +252,66 @@ def test_dragon4_interface(self):
+        a = np.float128('2')/np.float128('3')
+        b = np.float128(str(a))
+        assert_equal(str(a), str(b))
+        assert_(a != b)
```
Deliberately avoiding `assert_equal` because the message wouldn't be useful? Might be worth adding a comment about that here.

ahaldane replied:
It is because we don't have `assert_not_equal` available.

It is implemented in `np.ma.testutils`, but I don't want to import from the ma module.

eric-wieser commented on numpy/core/tests/test_scalarprint.py:
```diff
@@ -250,6 +252,66 @@ def test_dragon4_interface(self):
+            "0.0"]
+        assert_equal(got, expected)
+
+        # Note: we follow glibc behavior, but it (or gcc) might not be right.
```
Are we following the printing behavior, or the parsing behavior?

ahaldane replied:
The printing behavior, since the two values are different yet print the same. If we want to "fix" it, we need to account for the 107th bit in the dragon4 code above, by slightly changing the "shift" calculation, for example so that the end double's mantisa is stored in bits 54 to 0 intead of 53 to 0, and change 53 to 54 in some places.

I reported this issue here as a question:
https://sourceware.org/bugzilla/show_bug.cgi?id=22752

But note there is previous discussion I linked there about handling the 107th bit.

ahaldane replied:
Or, correction, it could be that gcc's float128 multiply operation isn't quite right.

I would summarize the situation as follows:  gcc accounts for the 107th bit when multiplying/dividing, but glibc ignores the 107th bit when printing.

eric-wieser commented:
It's pretty easy to get lost in this diff - could you spin off the first few maint commits into a second PR?

charris commented:
@ahaldane Do you have time to work on this in the near future?

ahaldane commented:
Yes, I have some time this weeked.

ahaldane commented:
All right, reorganized, rebased and updated based on comments.

I put all the commits except the ppc one in a new PR at #11167. So maybe we can merge that first, and then I can rebase so that we only end up with the signle float128 ppc commit here.

eric-wieser commented:
I think the scariest change here is the conversion of the mantissa from a uint64_t to a bigint. That was the one I was hoping to review by itself, but perhaps that's too much extra work to reorder the history

charris commented:
@ahaldane Needs rebase.

charris commented:
Let's put this in, for testing if nothing else. I don't much like the check for reentry, but suspect it will serve to indicate any future problems. The `dragon4.c` code can use more cleanups, so I don't think this will be the last time we look at this.

Thanks Allan.
